// ngx-dynamic-form/esm5/lib/directives/dynamic-add-field/dynamic-add-field.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, Input, ViewContainerRef, ComponentFactoryResolver } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { FormInputComponent } from '../../components/form-input/form-input.component';
import { FormButtonComponent } from '../../components/button/button.component';
import { FormSelectComponent } from '../../components/select/select.component';
import { FormFileUploadComponent } from '../../components/file-upload/file-upload.component';
import { FormDatetimeComponent } from '../../components/datetime/datetime.component';
/** @type {?} */
var components = {
    button: FormButtonComponent,
    input: FormInputComponent,
    select: FormSelectComponent,
    file: FormFileUploadComponent,
    date: FormDatetimeComponent
};
var DynamicAddFieldDirective = /** @class */ (function () {
    function DynamicAddFieldDirective(resolver, vcr) {
        this.resolver = resolver;
        this.vcr = vcr;
    }
    /**
     * @return {?}
     */
    DynamicAddFieldDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!components[this.config.type]) {
            /** @type {?} */
            var supportedTypes = Object.keys(components).join(', ');
            throw new Error("Trying to use an unsupported type (" + this.config.type + ").\n            Supported types: " + supportedTypes);
        }
        /** @type {?} */
        var component = components[this.config.type];
        /** @type {?} */
        var factory = this.resolver.resolveComponentFactory(component);
        this.component = this.vcr.createComponent(factory);
        this.component.instance.config = this.config;
        this.component.instance.formGroup = this.formGroup;
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    DynamicAddFieldDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.component) {
            this.component.instance.config = this.config;
            this.component.instance.formGroup = this.formGroup;
        }
    };
    DynamicAddFieldDirective.decorators = [
        { type: Directive, args: [{ selector: '[dynamicField]' },] },
    ];
    /** @nocollapse */
    DynamicAddFieldDirective.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ViewContainerRef }
    ]; };
    DynamicAddFieldDirective.propDecorators = {
        config: [{ type: Input }],
        formGroup: [{ type: Input }]
    };
    return DynamicAddFieldDirective;
}());
export { DynamicAddFieldDirective };
function DynamicAddFieldDirective_tsickle_Closure_declarations() {
    /** @type {?} */
    DynamicAddFieldDirective.prototype.config;
    /** @type {?} */
    DynamicAddFieldDirective.prototype.formGroup;
    /** @type {?} */
    DynamicAddFieldDirective.prototype.component;
    /** @type {?} */
    DynamicAddFieldDirective.prototype.resolver;
    /** @type {?} */
    DynamicAddFieldDirective.prototype.vcr;
}
