// ngx-dynamic-form/esm2015/lib/components/form-input/form-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding } from '@angular/core';
export class FormInputComponent {
    constructor() { }
    /**
     * @return {?}
     */
    get hostClass() {
        return this.config.divClass;
    }
}
FormInputComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'form-input',
                styles: [`input[type=submit]{margin-top:.5rem 1rem}.form-input{height:38px;border-radius:3.5px;background-color:#fff;border:1px solid #bdbdbd;margin-bottom:15px}label{margin-bottom:8px;font-size:16px}`, `input[type=submit]{margin-top:.5rem 1rem}.form-input{height:38px;border-radius:3.5px;background-color:#fff;border:1px solid #bdbdbd;margin-bottom:15px}label{margin-bottom:8px;font-size:16px}`],
                template: `<div *ngIf='!config.hidden' [formGroup]="formGroup">
    <label [ngClass]="config.labelClass" *ngIf="config.label && config.label!=''">{{ config.label }}</label>
    <input *ngIf="config.type ==='input' && !config.fromEnum " class="form-input" value={{config.text}} [disabled]="config.disabled" type={{config.inputType}} [ngClass]="config.class" [attr.placeholder]="config.placeholder" [formControlName]="config.name"
    />

    <select *ngIf="config.type==='select' && !config.fromEnum && !config.hidden" class="form-input" [formControlName]="config.name" [ngClass]=config.class>
        <option value='' [selected]="true">{{ config.placeholder }}</option>
        <option value={{option}}
                *ngFor="let option of config.options">
                {{option}}
        </option>
    </select>
    <select class="form-input" *ngIf="config.type==='select' && config.fromEnum && !config.hidden" [formControlName]="config.name" [ngClass]=config.class>
        <option value="" [selected]="true">{{ config.placeholder }}</option>
        <option value={{config.options[0][i]}}
                *ngFor="let o of config.options[0]; let i=index">
                {{ config.options[1][i] }}
        </option>
    </select>

    <textarea class="form-input" *ngIf="config.type==='textarea' && !config.hidden" [formControlName]="config.name" [attr.placeholder]="config.placeholder" rows={{config.rows}} cols={{config.cols}}></textarea>

    <ng-datepicker class="form-input" [ngClass]="config.class" *ngIf="config.type ==='date'" [formControlName]="config.name"></ng-datepicker>
</div>`
            },] },
];
/** @nocollapse */
FormInputComponent.ctorParameters = () => [];
FormInputComponent.propDecorators = {
    hostClass: [{ type: HostBinding, args: ['class',] }]
};
function FormInputComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormInputComponent.prototype.config;
    /** @type {?} */
    FormInputComponent.prototype.formGroup;
}
