// ngx-dynamic-form/esm2015/lib/components/file-upload/file-upload.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding } from '@angular/core';
export class FormFileUploadComponent {
    /**
     * @return {?}
     */
    get hostClass() {
        return this.config.divClass;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    handleFileUpload(event) {
        (/** @type {?} */ (this.formGroup.controls[this.config.name]))
            .setValue(this.config.multiple ? event.currentTarget.files : event.currentTarget.files[0]);
    }
}
FormFileUploadComponent.decorators = [
    { type: Component, args: [{
                selector: 'form-file-upload',
                styles: [``, `input[type=submit]{margin-top:.5rem 1rem}.form-input{height:38px;border-radius:3.5px;background-color:#fff;border:1px solid #bdbdbd;margin-bottom:15px}label{margin-bottom:8px;font-size:16px}`],
                template: `<div *ngIf='!config.hidden' [formGroup]="formGroup">
    <label [ngClass]="config.labelClass" *ngIf="config.label && config.label!=''">{{ config.label }}</label>
    <div class="dynamic-field form-select">
        <input [disabled]="config.disabled" class="form-input" type='file' (change)='handleFileUpload($event)' [accept]="config.accept" [ngClass]="config.class" [attr.placeholder]="config.placeholder" />
    </div>
</div>`
            },] },
];
FormFileUploadComponent.propDecorators = {
    hostClass: [{ type: HostBinding, args: ['class',] }]
};
function FormFileUploadComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormFileUploadComponent.prototype.config;
    /** @type {?} */
    FormFileUploadComponent.prototype.formGroup;
}
/**
 * @record
 */
export function FileConfig() { }
function FileConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    FileConfig.prototype.multiple;
    /** @type {?} */
    FileConfig.prototype.accept;
}
