// ngx-dynamic-form/esm5/lib/components/select/select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding } from '@angular/core';
var FormSelectComponent = /** @class */ (function () {
    function FormSelectComponent() {
    }
    Object.defineProperty(FormSelectComponent.prototype, "hostClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.divClass;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Converts and Enum Object to an Array
     * @param arg  Enum
     * @returns Array<Array<any>>
     */
    /**
     * Converts and Enum Object to an Array
     * @param {?} arg  Enum
     * @return {?} Array<Array<any>>
     */
    FormSelectComponent.prototype.enumToArray = /**
     * Converts and Enum Object to an Array
     * @param {?} arg  Enum
     * @return {?} Array<Array<any>>
     */
    function (arg) {
        /** @type {?} */
        var numKeyArr = [];
        /** @type {?} */
        var valArr = [];
        /** @type {?} */
        var returnedArray = [];
        /** @type {?} */
        var objKeys = Object.keys(arg);
        objKeys.forEach(function (key) {
            /** @type {?} */
            var numberKey = Number(key);
            if (!isNaN(numberKey)) {
                numKeyArr.push(numberKey);
            }
            else {
                valArr.push(key);
            }
            ;
        });
        returnedArray[0] = numKeyArr;
        returnedArray[1] = valArr;
        return returnedArray;
    };
    FormSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'form-select',
                    styles: ["", "input[type=submit]{margin-top:.5rem 1rem}.form-input{height:38px;border-radius:3.5px;background-color:#fff;border:1px solid #bdbdbd;margin-bottom:15px}label{margin-bottom:8px;font-size:16px}"],
                    template: "<div *ngIf='!config.hidden' [formGroup]=\"formGroup\">\n    <label [ngClass]=\"config.labelClass\" *ngIf=\"config.label && config.label!=''\">{{ config.label }}</label>\n    <div class=\"dynamic-field form-select\">\n        <select *ngIf=\"!config.enum\" class=\"form-input\" [hidden]='config.hidden' [formControlName]=\"config.name\" [ngClass]=config.class>\n                <option value='' [selected]=\"true\">{{ config.placeholder }}</option>\n                <option *ngFor=\"let option of config.options\"\n                        [value]='option'>\n                        {{option}}\n                </option>\n\n        </select>\n        <select [formControlName]=\"config.name\" class=\"form-input\" [ngClass]=config.class *ngIf=\"config.enum\">\n                <option value='' [selected]=\"true\">{{ config.placeholder }}</option>\n                <option *ngFor=\"let o of enumToArray(config.enum)[0]; let i=index\"\n                        [value]='enumToArray(config.enum)[0][i]'>\n                        {{ enumToArray(config.enum)[1][i] }}\n                </option>\n        </select>\n    </div>\n</div>"
                },] },
    ];
    FormSelectComponent.propDecorators = {
        hostClass: [{ type: HostBinding, args: ['class',] }]
    };
    return FormSelectComponent;
}());
export { FormSelectComponent };
function FormSelectComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormSelectComponent.prototype.config;
    /** @type {?} */
    FormSelectComponent.prototype.formGroup;
}
/**
 * @record
 */
function SelectConfig() { }
function SelectConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    SelectConfig.prototype.enum;
    /** @type {?|undefined} */
    SelectConfig.prototype.options;
}
