// ngx-dynamic-form/esm2015/lib/components/button/button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, IterableDiffers, HostBinding } from '@angular/core';
export class FormButtonComponent {
    /**
     * @param {?} differs
     */
    constructor(differs) {
        this.differ = differs.find([]).create(null);
    }
    /**
     * @return {?}
     */
    get hostClass() {
        return this.config.divClass;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        setTimeout(() => {
            /** @type {?} */
            const change = this.differ.diff(this.btn);
            if (document.querySelector('form').className.includes('ng-invalid')
                && this.config.inputType.toLowerCase() === 'submit') {
                this.config.disabled = true;
            }
            else {
                this.config.disabled = false;
            }
        }, 10);
    }
}
FormButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'form-button',
                styles: [``],
                template: `<div *ngIf='!config.hidden' class="form-input" [formGroup]="formGroup">
    <label [ngClass]="config.labelClass" *ngIf="config.label && config.label!=''">{{ config.label }}</label>
    <div class="dynamic-field form-button">
        <button [disabled]="config.disabled" [ngClass]="config.class" type="submit">
        {{ config.text }}
        </button>
    </div>
</div>`
            },] },
];
/** @nocollapse */
FormButtonComponent.ctorParameters = () => [
    { type: IterableDiffers }
];
FormButtonComponent.propDecorators = {
    hostClass: [{ type: HostBinding, args: ['class',] }]
};
function FormButtonComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormButtonComponent.prototype.config;
    /** @type {?} */
    FormButtonComponent.prototype.formGroup;
    /** @type {?} */
    FormButtonComponent.prototype.differ;
    /** @type {?} */
    FormButtonComponent.prototype.btn;
}
