// ngx-dynamic-form/esm2015/lib/components/datetime/datetime.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding } from '@angular/core';
export class FormDatetimeComponent {
    /**
     * @return {?}
     */
    get hostClass() {
        return this.config.divClass;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        (/** @type {?} */ (this.formGroup.controls[this.config.name]))
            .setValue(new Date(), { onlySelf: true });
    }
}
FormDatetimeComponent.decorators = [
    { type: Component, args: [{
                selector: 'form-datetime',
                styles: [``, `input[type=submit]{margin-top:.5rem 1rem}.form-input{height:38px;border-radius:3.5px;background-color:#fff;border:1px solid #bdbdbd;margin-bottom:15px}label{margin-bottom:8px;font-size:16px}`],
                template: `<div *ngIf='!config.hidden' [formGroup]="formGroup">
    <label *ngIf="config.label && config.label!=''">{{ config.label }}</label>
    <div class="dynamic-field form-datetime">
        <ng-datepicker class="form-input" [formControlName]="config.name"></ng-datepicker>

    </div>
</div>`
            },] },
];
FormDatetimeComponent.propDecorators = {
    hostClass: [{ type: HostBinding, args: ['class',] }]
};
function FormDatetimeComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormDatetimeComponent.prototype.config;
    /** @type {?} */
    FormDatetimeComponent.prototype.formGroup;
}
