// ngx-dynamic-form/esm2015/lib/components/select/select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding } from '@angular/core';
export class FormSelectComponent {
    /**
     * @return {?}
     */
    get hostClass() {
        return this.config.divClass;
    }
    /**
     * Converts and Enum Object to an Array
     * @param {?} arg  Enum
     * @return {?} Array<Array<any>>
     */
    enumToArray(arg) {
        /** @type {?} */
        const numKeyArr = [];
        /** @type {?} */
        const valArr = [];
        /** @type {?} */
        const returnedArray = [];
        /** @type {?} */
        const objKeys = Object.keys(arg);
        objKeys.forEach(key => {
            /** @type {?} */
            const numberKey = Number(key);
            if (!isNaN(numberKey)) {
                numKeyArr.push(numberKey);
            }
            else {
                valArr.push(key);
            }
            ;
        });
        returnedArray[0] = numKeyArr;
        returnedArray[1] = valArr;
        return returnedArray;
    }
}
FormSelectComponent.decorators = [
    { type: Component, args: [{
                selector: 'form-select',
                styles: [``, `input[type=submit]{margin-top:.5rem 1rem}.form-input{height:38px;border-radius:3.5px;background-color:#fff;border:1px solid #bdbdbd;margin-bottom:15px}label{margin-bottom:8px;font-size:16px}`],
                template: `<div *ngIf='!config.hidden' [formGroup]="formGroup">
    <label [ngClass]="config.labelClass" *ngIf="config.label && config.label!=''">{{ config.label }}</label>
    <div class="dynamic-field form-select">
        <select *ngIf="!config.enum" class="form-input" [hidden]='config.hidden' [formControlName]="config.name" [ngClass]=config.class>
                <option value='' [selected]="true">{{ config.placeholder }}</option>
                <option *ngFor="let option of config.options"
                        [value]='option'>
                        {{option}}
                </option>

        </select>
        <select [formControlName]="config.name" class="form-input" [ngClass]=config.class *ngIf="config.enum">
                <option value='' [selected]="true">{{ config.placeholder }}</option>
                <option *ngFor="let o of enumToArray(config.enum)[0]; let i=index"
                        [value]='enumToArray(config.enum)[0][i]'>
                        {{ enumToArray(config.enum)[1][i] }}
                </option>
        </select>
    </div>
</div>`
            },] },
];
FormSelectComponent.propDecorators = {
    hostClass: [{ type: HostBinding, args: ['class',] }]
};
function FormSelectComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormSelectComponent.prototype.config;
    /** @type {?} */
    FormSelectComponent.prototype.formGroup;
}
/**
 * @record
 */
function SelectConfig() { }
function SelectConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    SelectConfig.prototype.enum;
    /** @type {?|undefined} */
    SelectConfig.prototype.options;
}
