// ngx-dynamic-form/esm5/lib/components/datetime/datetime.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding } from '@angular/core';
var FormDatetimeComponent = /** @class */ (function () {
    function FormDatetimeComponent() {
    }
    Object.defineProperty(FormDatetimeComponent.prototype, "hostClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.divClass;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FormDatetimeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        (/** @type {?} */ (this.formGroup.controls[this.config.name]))
            .setValue(new Date(), { onlySelf: true });
    };
    FormDatetimeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'form-datetime',
                    styles: ["", "input[type=submit]{margin-top:.5rem 1rem}.form-input{height:38px;border-radius:3.5px;background-color:#fff;border:1px solid #bdbdbd;margin-bottom:15px}label{margin-bottom:8px;font-size:16px}"],
                    template: "<div *ngIf='!config.hidden' [formGroup]=\"formGroup\">\n    <label *ngIf=\"config.label && config.label!=''\">{{ config.label }}</label>\n    <div class=\"dynamic-field form-datetime\">\n        <ng-datepicker class=\"form-input\" [formControlName]=\"config.name\"></ng-datepicker>\n\n    </div>\n</div>"
                },] },
    ];
    FormDatetimeComponent.propDecorators = {
        hostClass: [{ type: HostBinding, args: ['class',] }]
    };
    return FormDatetimeComponent;
}());
export { FormDatetimeComponent };
function FormDatetimeComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormDatetimeComponent.prototype.config;
    /** @type {?} */
    FormDatetimeComponent.prototype.formGroup;
}
