// ngx-dynamic-form/esm5/lib/components/button/button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, IterableDiffers, HostBinding } from '@angular/core';
var FormButtonComponent = /** @class */ (function () {
    function FormButtonComponent(differs) {
        this.differ = differs.find([]).create(null);
    }
    Object.defineProperty(FormButtonComponent.prototype, "hostClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.divClass;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FormButtonComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    FormButtonComponent.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout(function () {
            /** @type {?} */
            var change = _this.differ.diff(_this.btn);
            if (document.querySelector('form').className.includes('ng-invalid')
                && _this.config.inputType.toLowerCase() === 'submit') {
                _this.config.disabled = true;
            }
            else {
                _this.config.disabled = false;
            }
        }, 10);
    };
    FormButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'form-button',
                    styles: [""],
                    template: "<div *ngIf='!config.hidden' class=\"form-input\" [formGroup]=\"formGroup\">\n    <label [ngClass]=\"config.labelClass\" *ngIf=\"config.label && config.label!=''\">{{ config.label }}</label>\n    <div class=\"dynamic-field form-button\">\n        <button [disabled]=\"config.disabled\" [ngClass]=\"config.class\" type=\"submit\">\n        {{ config.text }}\n        </button>\n    </div>\n</div>"
                },] },
    ];
    /** @nocollapse */
    FormButtonComponent.ctorParameters = function () { return [
        { type: IterableDiffers }
    ]; };
    FormButtonComponent.propDecorators = {
        hostClass: [{ type: HostBinding, args: ['class',] }]
    };
    return FormButtonComponent;
}());
export { FormButtonComponent };
function FormButtonComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    FormButtonComponent.prototype.config;
    /** @type {?} */
    FormButtonComponent.prototype.formGroup;
    /** @type {?} */
    FormButtonComponent.prototype.differ;
    /** @type {?} */
    FormButtonComponent.prototype.btn;
}
